### PR TITLE
Exec support using libpod api

### PR DESCRIPTION
**Pull request by towe75**
Alternative implementation to #85
It uses libpod api directly, following the minimal client strategy of our 0.2.0 release.

It's pretty complete and support both interactive use and scripted health checks. The only known issue is related to terminal resize events due to https://github.com/containers/podman/issues/7102 which should be fixed in the upcoming podman 3.0 release.

### CODE REVIEW

**Comment by towe75** [comment]
@drewbailey  PTAL. I think it's low risk to merge because it adds a totally new feature without changing any of the existing functions.

File: api/exec_create.go
Comment (drewbailey) [inline api/exec_create.go:50]:
```suggestion
// ExecSessionResponse contains the ID of a newly created exec session
```

File: api/exec_create.go
Comment (drewbailey) [inline api/exec_create.go:52]: Nit but I think ID is preferred 
```suggestion
	ID string
```

File: api/exec_start.go
Comment (drewbailey) [inline api/exec_start.go:172]: Can we return nil after the copy and drop this else

File: api/exec_start.go
Comment (drewbailey) [inline api/exec_start.go:101]: having our own clone function may be helpful here, but for now I think something like the following would be a bit safer
```
	client := new(http.Client)
	*client = *c.httpClient
```

File: api/exec_start.go
Comment (drewbailey) [inline api/exec_start.go:90]:
```suggestion
			c.logger.Trace("Resize terminal", "sessionId", sessionId, "height", size.Height, "width", size.Width)
```

File: driver.go
Comment (drewbailey) [inline driver.go:724]: let's handle these error cases

**Review by drewbailey — COMMENTED** [review]
I tried this out locally and it works great! Excellent work :tada: Just a few comments, Some tests would be good too

File: api/exec_start.go
Comment (towe75) [reply api/exec_start.go:172]: options.Tty is false for non interactive exec sessions, used by consul health checks. I tried to mimic what the docker driver does and the original podman code had this if/else case to support it.
We can try to simplify but it will require some testing

File: api/exec_start.go
Comment (towe75) [reply api/exec_start.go:101]: I'm fine with the change because i am not really a go expert ;-)

**Comment by towe75** [comment]
@drewbailey : i rebased this branch to include the latest changes from main. Please check my earlier reply on your review.

Are we good to go? I think it would be worth a release.

File: api/exec_start.go
Comment (drewbailey) [inline api/exec_start.go:88]:
```suggestion
			c.logger.Trace("Resize handler is done")
```

File: api/exec_start.go
Comment (drewbailey) [inline api/exec_start.go:94]:
```suggestion
				c.logger.Error("failed to resize TTY", "err", rerr)
```

**Comment by drewbailey** [comment]
@towe75 this looks really great! A few tests would be great to ensure everything works as expected as new podman versions roll out. There should be a few good examples in the Nomad repo

**Comment by towe75** [comment]
@drewbailey : thank you for feedback. I'll try to add some tests.

**Comment by drewbailey** [comment]
@towe75 We can also merge this and do a follow up PR with some tests before releasing if you prefer, I'll approve the PR and leave it up to you. If we can fix those two log levels that would be great

**Review by drewbailey — APPROVED** [review]

File: api/exec_start.go
Comment (towe75) [reply api/exec_start.go]: done

**Comment by towe75** [comment]
@drewbailey i started to derive a test from the other task driver and only two sub-tests from StreamingConformingTests fail:

--- FAIL: TestPodmanDriver_execTaskStreaming (76.20s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_notty:_basic (0.13s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_notty:_streaming (3.12s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_notty:_stty_check (0.11s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_notty:_stdin_passing (0.12s)
    --- FAIL: TestPodmanDriver_execTaskStreaming/basic:_notty:_children_processes (1.09s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_tty:_basic (0.13s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_tty:_streaming (3.07s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_tty:_stty_check (1.10s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_tty:_stdin_passing (0.11s)
    --- PASS: TestPodmanDriver_execTaskStreaming/basic:_tty:_children_processes (1.10s)
    --- FAIL: TestPodmanDriver_execTaskStreaming/isolation (0.28s)

notty:children_processes has to be tackled, could be potential bug. But "isolation" test has a hardcoded docker assumption in exec_testing.go . Should this become a bug in the nomad repo?

I did not commit the test yet to not break the build.

**Comment by drewbailey** [comment]
@towe75 that is great news! sounds like `TestExecFSIsolation` should pass in the expected freezer cgroup instead of assuming nomad or docker

**Comment by towe75** [comment]
I looked into the other failing test and it seems to be a podman problem OR wrong assumption. 
Here is what i tried:

```
podman run -d -t busybox --name tester

podman exec tester /bin/sh -c "(( sleep 3; echo from background ) & ); echo from main; exec sleep 1"
from main
```

A plain run without a container acts differently:

```
/bin/sh -c "(( sleep 3; echo from background ) & ); echo from main; exec sleep 1"
from main
from background
```

Should i raise a podman issue? Seems they handle it differently compared to docker.

**Comment by drewbailey** [comment]
@towe75 fyi we are removing this test https://github.com/hashicorp/nomad/pull/10291 from nomad

**Comment by towe75** [comment]
@drewbailey thank you for keeping me posted. I will file a follow-up issue to add the tests later because there is still the TestExecFSIsolation blocker. 
After that we can merge this PR, IMHO.

**Pull request merged**
